cocotb/regression.py

```diff
@@ -180,34 +185,41 @@ def _discover_tests() -> Iterable[Test]:
                 _logger.info(traceback.format_exc())
                 raise
 
-            if test_str:
-
+            if tests:
                 # Specific functions specified, don't auto-discover
-                for test_name in test_str.rsplit(','):
+                for test_name in tests:
                     try:
                         test = getattr(module, test_name)
                     except AttributeError:
-                        _logger.error("Requested test %s wasn't found in module %s", test_name, module_name)
-                        err = AttributeError("Test %s doesn't exist in %s" % (test_name, module_name))
-                        raise err from None  # discard nested traceback
-
-                    if not isinstance(test, Test):
-                        _logger.error("Requested %s from module %s isn't a cocotb.test decorated coroutine",
-                                      test_name, module_name)
-                        raise ImportError("Failed to find requested test %s" % test_name)
+                        not_found_tests.append(test_name)
+                    else:
+                        if not isinstance(test, Test):
+                            _logger.error("Requested %s from module %s isn't a cocotb.test decorated coroutine",
+                                          test_name, module_name)
+                            raise ImportError("Failed to find requested test %s" % test_name)
 
-                    # If we request a test manually, it should be run even if skip=True is set.
-                    test.skip = False
+                        # If we request a test manually, it should be run even if skip=True is set.
+                        test.skip = False
 
-                    yield test
+                        yield test
 
-                # only look in first module for all functions and don't complain if all functions are not found
-                break
+                # Clear not_found_tests for the next module search
+                tests = not_found_tests.copy()
+                not_found_tests.clear()
 
-            # auto-discover
-            for thing in vars(module).values():
-                if isinstance(thing, Test):
-                    yield thing
+                # Break if all tests were found
+                if not tests:
+                    break
```

Comment:
```suggestion
```
not needed with my change to `None` above

Reply:
I'm not sure if it is a good idea. After removing this `break`, we will always import all listed in the `MODULE` modules, even if all tests were found in the first one. So in cases where user has many modules and often runs some basic test that is in the first module, this can have a negative impact on performance.

Reply:
That is actually desired. Module-level code can have side effects. So if they are listed we should intend to load them, at a minimum.

Reply:
Ok, thanks for the explanation, I'll add this change.

Reply:
Another argument in favor of my suggestion is that it won't silently ignore missing modules just because it found all the tests already.

Reply:
Changed in ccc0ce20dd809c137e93ae20666fd06190fa0f4e.